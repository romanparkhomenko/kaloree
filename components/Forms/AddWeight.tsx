import React, { useState } from 'react';

import { Card, CardBody, CardTitle, Container, Row, Col, Form, FormGroup, Input } from 'reactstrap';
import Router, { useRouter } from 'next/router';

const AddWeight: React.FC = () => {
  const [pounds, setPounds] = useState('');
  const [successMessage, setSuccessMessage] = useState('');
  const [errorMessage, setErrorMessage] = useState('');

  const submitWeight = async (e: React.SyntheticEvent) => {
    e.preventDefault();
    setSuccessMessage('');
    setErrorMessage('');
    try {
      const body = { pounds };
      await fetch(`/api/weight`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      })
        .then(res => res.json())
        .then(response => {
          console.info(response);
          setSuccessMessage('Successfully logged weight');
          setPounds('');
          setTimeout(() => {
            setSuccessMessage('');
          }, 5000);
        })
        .catch(error => {
          setErrorMessage('Whoops, something went wrong');
          console.info(error);
        });
    } catch (error) {
      setErrorMessage('Whoops, something went wrong');
      console.error(error);
    }
  };

  return (
    <>
      <Card className="card-stats mb-4 mb-xl-0">
        <CardBody>
          <Row>
            <div className="col">
              <Form>
                <Col lg="12" className="align-items-center justify-content-between row">
                  <h6 className="heading-small text-muted mb-4">Log your Weight</h6>
                  <span className="icon icon-shape bg-gradient-primary text-white rounded-circle shadow">
                    <i className="ni ni-active-40" />
                  </span>
                </Col>
                <Row className="align-items-end">
                  <Col lg="12">
                    <label className="form-control-label" htmlFor="pounds">
                      Weight
                    </label>
                    <Input
                      className="form-control-alternative"
                      id="pounds"
                      placeholder="How much do you weigh?"
                      onChange={e => setPounds(e.target.value)}
                      value={pounds}
                      type="number"
                    />
                  </Col>
                </Row>
                <Row className="align-items-center justify-content-end">
                  <Col lg="12" className="mt-3">
                    <button
                      className="btn btn-primary bg-gradient-primary mb-1 w-100"
                      onClick={e => submitWeight(e)}
                    >
                      Log My Weight
                    </button>
                    {successMessage && (
                      <p className="text-center font-weight-700 text-info">{successMessage}</p>
                    )}
                    {errorMessage && (
                      <p className="text-center font-weight-700 text-warning">{errorMessage}</p>
                    )}
                  </Col>
                </Row>
              </Form>
            </div>
          </Row>
        </CardBody>
      </Card>
    </>
  );
};

export default AddWeight;
