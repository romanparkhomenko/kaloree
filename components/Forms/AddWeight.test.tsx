import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { render, fireEvent, waitFor, screen, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import AddWeight from './AddWeight';

describe('Add Weight Form', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<AddWeight />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('All inputs work and user can submit data', async () => {
    render(<AddWeight />);

    const weightInput = screen.getByPlaceholderText('How much do you weigh?');
    const submitButton = screen.getByText('Log My Weight');

    fireEvent.change(weightInput, { target: { value: '200' } });
    expect(weightInput).toHaveValue(200);

    const fakeWeightResponse = { okay: 'weight logged' };

    global.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        json: () => Promise.resolve(fakeWeightResponse),
      });
    });

    await act(async () => {
      fireEvent.click(submitButton);
    });

    // Confirm success message exists and values are reset
    expect(screen.getByText('Successfully logged weight')).toBeVisible();
    expect(weightInput).toHaveValue(null);
  });
});
