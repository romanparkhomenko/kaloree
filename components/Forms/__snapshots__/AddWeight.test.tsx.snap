// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Weight Form renders correctly 1`] = `
<div
  className="card-stats mb-4 mb-xl-0 card"
>
  <div
    className="card-body"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <form
          className=""
        >
          <div
            className="align-items-center justify-content-between row col-lg-12"
          >
            <h6
              className="heading-small text-muted mb-4"
            >
              Log your Weight
            </h6>
            <span
              className="icon icon-shape bg-gradient-primary text-white rounded-circle shadow"
            >
              <i
                className="ni ni-active-40"
              />
            </span>
          </div>
          <div
            className="align-items-end row"
          >
            <div
              className="col-lg-12"
            >
              <label
                className="form-control-label"
                htmlFor="pounds"
              >
                Weight
              </label>
              <input
                className="form-control-alternative form-control"
                id="pounds"
                onChange={[Function]}
                placeholder="How much do you weigh?"
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            className="align-items-center justify-content-end row"
          >
            <div
              className="mt-3 col-lg-12"
            >
              <button
                className="btn btn-primary bg-gradient-primary mb-1 w-100"
                onClick={[Function]}
              >
                Log My Weight
              </button>
              
              
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
